### PR TITLE
SolverRewards 3: Updated db schema and read/write functions

### PR DESCRIPTION
**Pull request by sunce86**
Autopilot and orderbook callsite functions will be introduced in a follow up PR.

### CODE REVIEW

File: database/sql/V048__create_settlement_rewards.sql [hunk 1]
```diff
@@ -12,20 +14,20 @@ CREATE TABLE settlement_observations (
...
-  gas_used numeric(78,0) NOT NULL,
-  effective_gas_price numeric(78,0) NOT NULL,
+  gas_used numeric(78,0),
+  effective_gas_price numeric(78,0),
   -- the surplus observed from the transaction call data,
   -- and converted to ETH with the auction external prices.
-  surplus numeric(78,0) NOT NULL,
-  fee numeric(78,0) NOT NULL,
+  surplus numeric(78,0),
+  fee numeric(78,0),
```
Comment (sunce86) [inline database/sql/V048__create_settlement_rewards.sql:23]: Removed `NOT NULL` because the table is initially populated (block_number and log_index) when GPv2Settlement event is emitted, while the rest is populated in an autopilot background task similar to `AuctionTransactionUpdater`.

File: database/sql/V048__create_settlement_rewards.sql [hunk 2]
```diff
@@ -12,20 +14,20 @@ CREATE TABLE settlement_observations (
...
 );
 
 CREATE TABLE auction_prices (
-  auction_id bigint NOT NULL,
-  token bytea NOT NULL,
-  price numeric(78,0) NOT NULL
+  auction_id bigint NOT NULL PRIMARY KEY,
+  tokens bytea[] NOT NULL,
+  prices numeric(78,0)[] NOT NULL
 );
```
Comment (sunce86) [inline database/sql/V048__create_settlement_rewards.sql:31]: Might not be the perfect design but seemed convenient because we always fetch all prices for an auction.

File: crates/database/src/events.rs [hunk 3]
```diff
@@ -64,7 +65,8 @@ pub async fn delete(
     ex.execute(sqlx::query(QUERY_TRADE).bind(delete_from_block_number))
         .await?;
 
-    const QUERY_SETTLEMENTS: &str = "DELETE FROM settlements WHERE block_number >= $1;";
+    const QUERY_SETTLEMENTS: &str = "WITH S AS (DELETE FROM settlements WHERE block_number >= $1) \
+                                     DELETE FROM settlement_observations WHERE block_number >= $1;";
```
Comment (MartinquaXD) [inline crates/database/src/events.rs:69]: Here you could do 2 queries without needing to refactor anything as you already have a `PgTransaction` available.

File: database/sql/V048__create_settlement_rewards.sql [hunk 4]
```diff
@@ -3,7 +3,9 @@
 CREATE TABLE settlement_scores (
   auction_id bigint PRIMARY KEY,
   winning_score numeric(78,0) NOT NULL,
-  reference_score numeric(78,0) NOT NULL
+  reference_score numeric(78,0) NOT NULL,
```
Comment (MartinquaXD) [inline database/sql/V048__create_settlement_rewards.sql:6]: What happens if only a single solver offers a valid solution?
Should we maybe keep `reference_score` nullable for that case instead of using a default value?
Also a comment on what the `reference_score` is would be nice.

File: database/sql/V048__create_settlement_rewards.sql (hunk 2 above)
Comment (nlordell) [reply database/sql/V048__create_settlement_rewards.sql:31]: Hmm, I think we should have a table with a single row per price. Then, we would insert multiple rows per auction. I am not a fan of using arrays like this in this table.

File: database/sql/V048__create_settlement_rewards.sql [hunk 5]
```diff
@@ -12,20 +14,20 @@ CREATE TABLE settlement_observations (
...
 CREATE TABLE auction_prices (
-  auction_id bigint NOT NULL,
-  token bytea NOT NULL,
-  price numeric(78,0) NOT NULL
+  auction_id bigint NOT NULL PRIMARY KEY,
+  tokens bytea[] NOT NULL,
+  prices numeric(78,0)[] NOT NULL
 );
 
 CREATE TABLE auction_participants (
```
Comment (nlordell) [inline database/sql/V048__create_settlement_rewards.sql:33]: I missed the PR that added this, but I also think that we should have 1 row per participant and not use arrays.

At least this is my simple understanding of how this "should" be (and allow you to query things like `SELECT COUNT(*) from auction_participants where participant = '\x...'` for querying the number of times an address participated in an auction).

File: crates/database/src/auction_prices.rs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+pub struct Prices {
+    pub tokens: Vec<Address>,
+    pub prices: Vec<BigDecimal>,
+}
+
+pub async fn upsert(
+    ex: &mut PgConnection,
+    auction_id: AuctionId,
+    tokens: Vec<Address>,
+    prices: Vec<BigDecimal>,
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/database/src/auction_prices.rs:18]: Sure, but then I need to change the `fetch` functions also. But I think that's ok also.

File: crates/database/src/events.rs (hunk 3 above)
Comment (sunce86) [reply crates/database/src/events.rs:69]: With this I was aiming to get the atomic execution because my old code assumed that `settlements` and `settlement_observations` tables are in sync always. But I think it is ok to decouple them.

File: crates/database/src/events.rs [hunk 7]
```diff
@@ -139,15 +141,18 @@ async fn insert_settlement(
     event: &Settlement,
 ) -> Result<(), sqlx::Error> {
     const QUERY: &str = "\
-        INSERT INTO settlements (tx_hash, block_number, log_index, solver) VALUES ($1, $2, $3, $4) \
-                         ON CONFLICT DO NOTHING;";
+    WITH S AS (INSERT INTO settlements (tx_hash, block_number, log_index, solver) VALUES ($1, $2, \
+                         $3, $4) ON CONFLICT DO NOTHING)
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/database/src/events.rs:145]: Yes, I did not find other way. But this can be decoupled also, because they can be updated separately.

File: crates/database/src/auction_participants.rs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+#[derive(Debug, PartialEq, sqlx::FromRow)]
+pub struct Participants {
+    pub auction_id: AuctionId,
+    pub participants: Vec<Address>,
+}
+
+pub async fn insert(
+    ex: &mut PgConnection,
+    auction_id: AuctionId,
+    participants: Vec<Address>,
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/database/src/auction_participants.rs]: Fixed.

File: crates/database/src/auction_transaction.rs [hunk 9]
```diff
@@ -70,6 +70,26 @@ LIMIT 1
         .await
 }
 
+#[derive(Debug, sqlx::FromRow)]
+pub struct Auction {
+    pub auction_id: AuctionId,
+}
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/database/src/auction_transaction.rs]: Cool, fixed.

File: crates/database/src/settlement_scores.rs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    pub reference_score: BigDecimal,
+    pub block_deadline: i64,
+}
+
+pub async fn insert(
+    ex: &mut PgConnection,
+    auction_id: AuctionId,
+    winning_score: BigDecimal,
+    reference_score: BigDecimal,
+    block_deadline: i64,
```
Comment (bh2smith) [inline crates/database/src/settlement_scores.rs:16]: Don't we want to just pass the score struct into this method and unpack it inside the function with dot notation?

File: crates/database/src/settlements.rs [hunk 11]
```diff
@@ -18,6 +18,25 @@ WHERE
         .await
 }
 
+pub async fn get_hash_by_event(
```
Comment (bh2smith) [inline crates/database/src/settlements.rs:21]: I find the naming a bit awkward here because there is no "event" being passed. If it were actually an Event we and we unpacked the (block_number, log_index) inside the method I wouldn't have an issue. 

Also, I think this should return and Optional Hash (because for some block_numbers there may not be a SettlementEvent). It is pretty clear this function is meant to be for a SettlementEvent (since the raw sql is querying the settlement table).

File: database/sql/V048__create_settlement_rewards.sql (hunk 1 above)
Comment (bh2smith) [reply database/sql/V048__create_settlement_rewards.sql:23]: It was my impression that this table will not be populated when no corresponding transaction exists. This is why we had set these fields to NOT NULL. If this table is going to be populated with vacant data. I may have to change the data export queries to match this.

File: database/sql/V048__create_settlement_rewards.sql (hunk 5 above)
Comment (bh2smith) [reply database/sql/V048__create_settlement_rewards.sql:33]: I would be very much in favour of not using arrays. They are hard to parse and also make transposition a two-step process!

**Review by bh2smith — APPROVED** [review]
I like what I see here. There are a few places where some things could be changed. Assuming that most of the key concerns are addressed, I think its close enough to approve.

In particular there are a few places where the structure attributes are passed into methods rather than the structure itself. This leaves the function signature unbound to the object that it is clearly bound to operate on (based on the internal implementation). Should definitely be fixed.

File: crates/database/src/auction_prices.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (MartinquaXD) [reply crates/database/src/auction_prices.rs:18]: I initially thought so, too. But it's indeed true that liquidity orders and limit orders can result in multiple prices for the same tokens. So having 2 lists here should be correct.

File: database/sql/V048__create_settlement_rewards.sql (hunk 1 above)
Comment (sunce86) [reply database/sql/V048__create_settlement_rewards.sql:24]: It will be populated with the vacant data only initially when our system detects mined transaction.
Then, in the background task it will be populated with full data following the FINALIZED_BLOCK rule, meaning it will be lagging for 64 blocks. 
Since you are doing your queries on a historic data I think there is no need to change them.

File: database/sql/V048__create_settlement_rewards.sql (hunk 2 above)
Comment (sunce86) [reply database/sql/V048__create_settlement_rewards.sql]: Reverted.

File: crates/database/src/auction_prices.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/database/src/auction_prices.rs]: Refactored. Can you check again pls.

File: database/sql/V048__create_settlement_rewards.sql (hunk 4 above)
Comment (sunce86) [reply database/sql/V048__create_settlement_rewards.sql:7]: Added comment for reference_score.

File: database/sql/V048__create_settlement_rewards.sql (hunk 5 above)
Comment (sunce86) [reply database/sql/V048__create_settlement_rewards.sql]: Fixed.

File: crates/database/src/events.rs [hunk 12]
```diff
@@ -47,6 +47,7 @@ pub async fn last_block(ex: &mut PgConnection) -> Result<i64, sqlx::Error> {
     const QUERY: &str = "\
             SELECT GREATEST( (SELECT COALESCE(MAX(block_number), 0) FROM trades), (SELECT \
                          COALESCE(MAX(block_number), 0) FROM settlements), (SELECT \
+                         COALESCE(MAX(block_number), 0) FROM settlement_observations),(SELECT \
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/database/src/events.rs:50]: I added it for completeness, it does not hurt. However, the function itself is not reliable since the tables can be left in an inconsistent state (append and delete functions are not atomic). At least this is my understanding.

File: crates/database/src/settlement_observations.rs [hunk 13]
```diff
@@ -0,0 +1,164 @@
+use {bigdecimal::BigDecimal, sqlx::PgConnection};
+
+#[derive(Debug, PartialEq, sqlx::FromRow)]
+pub struct SettlementEvent {
+    pub block_number: i64,
+    pub log_index: i64,
+}
```
[the remark this answers is not in the document]
Comment (sunce86) [reply crates/database/src/settlement_observations.rs]: Yes, reused `EventIndex` now.

File: crates/database/src/settlement_scores.rs (hunk 10 above)
Comment (sunce86) [reply crates/database/src/settlement_scores.rs]: Done.

File: crates/database/src/auction_prices.rs [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+/// External token price for a given auction.
+#[derive(Debug, Clone, PartialEq, sqlx::FromRow)]
+pub struct AuctionPrice {
+    pub auction_id: AuctionId,
+    pub token: Address,
+    pub price: BigDecimal,
+}
+
+pub async fn insert(ex: &mut PgConnection, prices: Vec<AuctionPrice>) -> Result<(), sqlx::Error> {
```
Comment (MartinquaXD) [inline crates/database/src/auction_prices.rs:15]: I think it would probably be best to use a DB transaction here. Otherwise half of the inserts might fail and we assume everything is in order when looking at the DB.

File: crates/database/src/settlements.rs (hunk 11 above)
Comment (sunce86) [reply crates/database/src/settlements.rs:21]: As far as I see, `settlements` table is always fully populated, so no need to query optional by hash.
There is also no `by_block_number` query here.
I fixed the event comment.

File: crates/database/src/auction_prices.rs (hunk 14 above)
Comment (MartinquaXD) [reply crates/database/src/auction_prices.rs:15]: I think you only have to a transaction:
```rust 
pub async fn insert(ex: &mut PgConnection, prices: Vec<AuctionPrice>) -> Result<(), sqlx::Error> {
    let mut ex = ex.begin().await?;
    // existing insert logic
    ex.commit().await
}
```

Comment (MartinquaXD) [reply crates/database/src/auction_prices.rs:15]: The same pattern can be applied across multiple functions by passing in a `&mut PgTransaction<'_>` to each one. This would be nicer in case we want to insert multiple things in different tables at once. (Which might be needed for the new logic)

File: crates/database/src/auction_participants.rs [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+/// Participants of a solver competition for a given auction.
+#[derive(Debug, Clone, PartialEq, sqlx::FromRow)]
+pub struct Participants {
+    pub auction_id: AuctionId,
+    pub participants: Vec<Address>,
+}
+
+pub async fn insert(ex: &mut PgConnection, data: Participants) -> Result<(), sqlx::Error> {
+    const QUERY: &str =
+        r#"INSERT INTO auction_participants (auction_id, participants) VALUES ($1, $2);"#;
```
Comment (nlordell) [inline crates/database/src/auction_participants.rs:15]: I don't think participants should be an array here.

File: crates/database/src/auction_prices.rs (hunk 14 above)
Comment (nlordell) [reply crates/database/src/auction_prices.rs:15]: I also think we should follow the structure of the `database` crate and pass in a `PgTransaction` here. Kind of what we do for other queries:

https://github.com/cowprotocol/services/blob/43a7f9cb9271c47d7c1cce6b2cda72286e106351/crates/database/src/events.rs#L78-L95

Comment (sunce86) [reply crates/database/src/auction_prices.rs:15]: Fixed.

File: crates/database/src/settlement_observations.rs [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+
+pub async fn update(
+    ex: &mut PgConnection,
+    block_number: i64,
+    log_index: i64,
+    gas_used: BigDecimal,
+    effective_gas_price: BigDecimal,
+    surplus: BigDecimal,
+    fee: BigDecimal,
```
Comment (nlordell) [inline crates/database/src/settlement_observations.rs:36]: Might be nice to group this into an `SettlementObservation` struct.

File: database/sql/V048__create_settlement_rewards.sql [hunk 17]
```diff
@@ -12,12 +15,12 @@ CREATE TABLE settlement_observations (
   -- table, read from the transaction receipt
   block_number bigint NOT NULL,
   log_index bigint NOT NULL,
-  gas_used numeric(78,0) NOT NULL,
-  effective_gas_price numeric(78,0) NOT NULL,
+  gas_used numeric(78,0),
+  effective_gas_price numeric(78,0),
```
Comment (nlordell) [inline database/sql/V048__create_settlement_rewards.sql:19]: In my mind, these need to be nullable.

Specifically, the `driver`/`solver` needs to ask to populate this **before** mining the transaction, when `gas_used` and `effective_gas_price` is unknown.

Comment (sunce86) [reply database/sql/V048__create_settlement_rewards.sql:19]: Agree they should be nullable, but I don't understand why it needs to be populated before mining the transaction? I insert it for the first time when the `settlement event` is raised - transaction mined.

**Review by MartinquaXD — APPROVED** [review]
My suggestions/questions have been addressed. We might have to revisit the DB transaction discussion when we see how those functions actually get used, though. For right now it seems fine to me.

File: crates/database/src/auction_participants.rs (hunk 15 above)
Comment (sunce86) [reply crates/database/src/auction_participants.rs]: Fixed.

File: crates/database/src/settlement_observations.rs (hunk 16 above)
Comment (sunce86) [reply crates/database/src/settlement_observations.rs]: Fixed.

File: crates/database/src/auction_participants.rs [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+/// Participant of a solver competition for a given auction.
+#[derive(Debug, Clone, PartialEq, sqlx::FromRow)]
+pub struct Participant {
+    pub auction_id: AuctionId,
+    pub participant: Address,
+}
+
+pub async fn insert(
+    ex: &mut PgConnection,
+    participants: Vec<Participant>,
```
Comment (vkgnosis) [inline crates/database/src/auction_participants.rs:15]:
```suggestion
    participants: &[Participant],
```

File: crates/database/src/auction_participants.rs [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+) -> Result<(), sqlx::Error> {
+    let mut transaction = ex.begin().await?;
```
Comment (vkgnosis) [inline crates/database/src/auction_participants.rs:17]: You shouldn't start a transaction inside of these functions usually. Instead change the type of `ex` to `&mut PgTransaction` and create the transaction in the caller. See `crates/database/src/lib.rs` for more information.

File: database/sql/V048__create_settlement_rewards.sql [hunk 20]
```diff
@@ -30,7 +33,7 @@ CREATE TABLE auction_prices (
 
 CREATE TABLE auction_participants (
  -- This links to the `auctions` table
- auction_id bigint PRIMARY KEY,
- -- All solvers who submitted a valid solution to the auction.
- participants bytea[]
+ auction_id bigint NOT NULL,
+ -- Solver who submitted a valid solution to the auction.
+ participant bytea NOT NULL
 );
```
Comment (vkgnosis) [inline database/sql/V048__create_settlement_rewards.sql:39]: This table has no  primary key but you are running queries on it of the form `SELECT * FROM auction_participants WHERE auction_id = $1`, which indicates you want an index on at least `auction_id`. Since it doesn't make sense for this table to contain the same `(auction_id, participant)` pair twice, we should probably make `(auction_id, participant) ` the primary key.

**Comment by sunce86** [comment]
I think we sorted out all open questions, merging.

File: crates/database/src/auction_prices.rs [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+            .bind(price.price)
+            .execute(&mut *transaction)
+            .await?;
+    }
+    transaction.commit().await?;
+    Ok(())
+}
+
+pub async fn delete(ex: &mut PgConnection, auction_id: AuctionId) -> Result<(), sqlx::Error> {
+    const QUERY: &str = "DELETE FROM auction_prices WHERE auction_id = $1";
```
Comment (vkgnosis) [inline crates/database/src/auction_prices.rs:32]: This table is missing a primary key on `(auction_id, token)` because it does not make sense to have duplicates and it is important for the efficiency of the query.

File: crates/database/src/events.rs (hunk 12 above)
Comment (vkgnosis) [inline crates/database/src/events.rs:50]: It doesn't make sense to me that we're handling the settlement_observations table in this function. This function shouldn't touch the table.Settlement events go into the `settlements` table. Why do we need to also put them in another table? This is related to another comment where you said you had to change some fields to be non null. I don't think this is the case. You should query the settlements table joined with the observations table in `get_settlement_event_without_observation` instead.

`auction_transaction` is designed to not have coumns be null because rows only get inserted when we have the data. This PR should work the same way.

File: crates/database/src/settlement_observations.rs [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+
+pub async fn get_settlement_event_without_observation(
+    ex: &mut PgConnection,
+    max_block_number: i64,
+) -> Result<Option<EventIndex>, sqlx::Error> {
+    const QUERY: &str = r#"
+SELECT block_number, log_index
+FROM settlement_observations
+WHERE gas_used IS NULL AND block_number <= $1
+LIMIT 1
```
Comment (vkgnosis) [inline crates/database/src/settlement_observations.rs:11]: I don't know how this function is going to be used but it seems like you need an index on `(block_number, gas_used)` in order to make this query efficient.

**Review by vkgnosis — CHANGES_REQUESTED** [review]

File: database/sql/V048__create_settlement_rewards.sql (hunk 1 above)
Comment (vkgnosis) [reply database/sql/V048__create_settlement_rewards.sql:24]: This isn't the right approach. See my review below.

File: crates/database/src/auction_participants.rs (hunk 18 above)
Comment (sunce86) [reply crates/database/src/auction_participants.rs:15]: Fixed in https://github.com/cowprotocol/services/pull/1270

File: crates/database/src/auction_participants.rs (hunk 19 above)
Comment (sunce86) [reply crates/database/src/auction_participants.rs:17]: Fixed in https://github.com/cowprotocol/services/pull/1270

File: database/sql/V048__create_settlement_rewards.sql (hunk 20 above)
Comment (sunce86) [reply database/sql/V048__create_settlement_rewards.sql:39]: Fixed in https://github.com/cowprotocol/services/pull/1270

Comment (bh2smith) [reply database/sql/V048__create_settlement_rewards.sql:39]: How are we supposed to determine the winning participant from this now? Used to the the last entry in the list and we still need to know this info. Specifically for the scenario when the auction is not settled and we want to know who to penalize.

File: crates/database/src/events.rs (hunk 12 above)
Comment (bh2smith) [reply crates/database/src/events.rs:50]: I also mentioned that this shouldn't be here but for slightly different reasons.

File: crates/database/src/settlements.rs (hunk 11 above)
Comment (bh2smith) [reply crates/database/src/settlements.rs:25]: Cool! If your passing a struct that is known to correspond to an entry in the table and not a random pair of integers then the return value does not need to be optional.

File: database/sql/V048__create_settlement_rewards.sql (hunk 20 above)
Comment (vkgnosis) [reply database/sql/V048__create_settlement_rewards.sql:39]: In another comment somewhere here we changed the type from an sql array to store one participant per row. This change removed all ordering of participants per auction. This is unrelated to my comment here. Postgres doesn't guarantee any order if there is no ordering constraint in the query.

File: crates/database/src/auction_prices.rs (hunk 21 above)
Comment (sunce86) [reply crates/database/src/auction_prices.rs:32]: Addressed in https://github.com/cowprotocol/services/pull/1271

File: crates/database/src/events.rs (hunk 12 above)
Comment (sunce86) [reply crates/database/src/events.rs:50]: Addressed in https://github.com/cowprotocol/services/pull/1271

File: crates/database/src/settlement_observations.rs (hunk 22 above)
Comment (sunce86) [reply crates/database/src/settlement_observations.rs:11]: Function not used anymore.
